### PR TITLE
massive cleanup of scope analysis

### PR DESCRIPTION
**Pull request by o-**

### CODE REVIEW

**Comment by o-** [comment]
this is my new iteration of the dataflow analysis. I think it got quite a bit simpler. it's still complex...

File: rir/src/compiler/analysis/abstract_value.h [hunk 1]
```diff
@@ -230,6 +230,11 @@ struct AbstractLoad {
     Value* env;
     AbstractPirValue result;
 
+    explicit AbstractLoad(const AbstractPirValue& val)
+        : env(AbstractREnvironment::UnknownParent), result(val) {
+        assert(env);
```
Comment (charig) [inline rir/src/compiler/analysis/abstract_value.h:235]: Why the assert?

File: rir/src/compiler/analysis/abstract_value.h [hunk 2]
```diff
@@ -263,16 +268,26 @@ class AbstractREnvironmentHierarchy {
         return res;
     }
 
+    bool known(Value* env) const { return envs.count(env); }
+
+    const AbstractREnvironment& at(Value* env) const {
```
Comment (charig) [inline rir/src/compiler/analysis/abstract_value.h:273]: Why we duplicate [] and at?

File: rir/src/compiler/pir/instruction.h [hunk 3]
```diff
@@ -86,13 +86,22 @@ enum class EnvAccess : uint8_t {
 };
 
 // Effect that can be produced by an instruction.
+// This is a trivial lattice, any effect with higher order contains all the
+// lower order effects.
 enum class Effect : uint8_t {
     None,
+    // Instruction doesn't really have effects itself, but it should not be
+    // hoisted over any other instruction with effect. Example: Assume
+    EffectOrderObserved,
```
Comment (charig) [inline rir/src/compiler/pir/instruction.h:95]: Nice! 
Maybe named it sth like `DependsOnLastEffect` / `NeedsLastEffect`

File: rir/src/compiler/analysis/abstract_value.cpp [hunk 4]
```diff
@@ -136,6 +137,41 @@ AbstractLoad AbstractREnvironmentHierarchy::get(Value* env, SEXP e) const {
...
 
+// Looking up functions is slightly trickier, since non-function bindings have
+// to be skipped.
+AbstractLoad AbstractREnvironmentHierarchy::getFun(Value* env, SEXP e) const {
+    assert(env);
+    if (aliases.count(env))
+        env = aliases.at(env);
+    while (env != AbstractREnvironment::UnknownParent) {
+        if (envs.count(env) == 0)
+            return AbstractLoad(env ? env : AbstractREnvironment::UnknownParent,
```
Comment (charig) [inline rir/src/compiler/analysis/abstract_value.cpp:148]: when does env could not be true

File: rir/src/compiler/analysis/abstract_value.cpp [hunk 5]
```diff
@@ -136,6 +137,41 @@ AbstractLoad AbstractREnvironmentHierarchy::get(Value* env, SEXP e) const {
...
+
+            // If it might be a closure, we can neither be sure, nor exclude
+            // this binding...
+            if (res.type.maybe(RType::closure))
+                return AbstractLoad(AbstractREnvironment::UnknownParent,
+                                    AbstractPirValue::tainted());
+        }
+        auto parent = envs.at(env).parentEnv();
+        assert(parent);
+        if (parent == AbstractREnvironment::UnknownParent &&
```
Comment (charig) [inline rir/src/compiler/analysis/abstract_value.cpp:166]: I do not get this if

File: rir/src/compiler/analysis/scope.h [hunk 6]
```diff
@@ -18,19 +18,178 @@ namespace pir {
...
+class ScopeAnalysisState {
+    AbstractREnvironmentHierarchy envs;
+    std::unordered_map<Instruction*, AbstractPirValue> returnValues;
+    AbstractPirValue result;
+    std::set<Instruction*> observedStores;
+    std::set<Value*> allStoresObserved;
+
+    bool mayUseReflection = false;
+
+    AbstractResult mergeCall(Code* cur, const ScopeAnalysisState& other) {
```
Comment (charig) [inline rir/src/compiler/analysis/scope.h:30]: Not sure why mergeCall does not need to merge the return values, but still, would be good to factorize here.

File: rir/src/compiler/opt/scope_resolution.cpp [hunk 7]
```diff
@@ -23,155 +23,209 @@ class TheScopeResolution {
...
-                    auto res = analysis.returnValues.at(i);
-                    if (res.isSingleValue() &&
-                        res.singleValue().origin->bb()->owner == function &&
-                        doms.dominates(res.singleValue().origin->bb(), bb)) {
-                        i->replaceUsesWith(res.singleValue().val);
+
+                auto before = analysis.at<ScopeAnalysis::BeforeInstruction>(i);
+                auto after = analysis.at<ScopeAnalysis::AfterInstruction>(i);
+
+                // Force can only see the callee env through reflection
```
Comment (charig) [inline rir/src/compiler/opt/scope_resolution.cpp:38]: callee or caller?

File: rir/src/compiler/opt/scope_resolution.cpp [hunk 8]
```diff
@@ -23,155 +23,209 @@ class TheScopeResolution {
...
-                            auto r =
-                                new StVar(ss->varName, ss->val(), aload.env);
-                            bb->replace(ip, r);
-                            ss->replaceUsesWith(r);
-                        }
+                    ip = next;
+                    continue;
+                }
+
+                analysis.lookup(after, i, [&](const AbstractLoad& aLoad) {
```
Comment (charig) [inline rir/src/compiler/opt/scope_resolution.cpp:69]: lookup?

**Review by charig — APPROVED** [review]

**Pull request merged**